### PR TITLE
o/snapstate: update default provider if missing required content

### PR DESCRIPTION
**Pull request by MiguelPires**
Updates a default provider when installing a snap if it's already installed but missing a required content attribute.

### CODE REVIEW

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:299]: I *think* we can hit store.ErrNoUpdateAvailable here if for whatever reason the content provider hasn't been updated in the store? Is this error going to surface and prevent entire install/update (would be change that breaks compatibility)?

Comment (MiguelPires) [reply overlord/snapstate/handlers.go:299]: Yes, that would prevent the installation. I incorrectly assumed that previously the installation would fail if the plug couldn't be auto-connected but I see that's not true (which makes sense). I can change it to log that snapd failed to update the prereq and proceed with the install

**Comment by stolowski** [comment]
@MiguelPires the unit test failure might be related to the changes?

**Comment by MiguelPires** [comment]
It is, the test is flaky because prerequisites are now stored in a map so the iteration order varies which makes the test check fail sometimes. I was waiting for https://github.com/snapcore/snapd/pull/10643 to fix it. I'll do it today

**Review by pedronis — DISMISSED** [review]
we need code to deal with conflicts/ongoing operations on the snap that needs updating. On top of that we also need an explicit test about a situation where the snap to update is also part of the same update change, we need a test equivalent to:

snap refresh consumer-with-new-content-need outdated-producer

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:363]: afaict this could reuse linkSnapInFlight  tweaking it a bit (mainly to return the relevant change), instead of defining a full new helper

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:350]: as we do for the Install path, this needs to check whether the error is a conflict and return a Retry in that case

**Review by pedronis — COMMENTED** [review]
the code seems to be going the right directions but some I think some reuse is possible and still missing some case handling, also there is a code conflict atm

**Comment by MiguelPires** [comment]
Thanks for the feedback @pedronis, I addressed those issues and added a test for a conflicting task from a different change. 
(I see you already self-requested a review, thanks!)

File: overlord/snapstate/conflict.go
Comment (pedronis) [inline overlord/snapstate/conflict.go:171]: this is equivalent to task.Status().Ready()

the change seems ok to me but probably needs to be done as a separate PR with some associated tests

File: overlord/snapstate/conflict.go
Comment (pedronis) [inline overlord/snapstate/conflict.go:173]: I would move this though just before the affectedSnaps code below

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:364]: I would leave out Reason as Install code does

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:218]: because of the conflict checks there shouldn't really be more than one non-ready link-snap task around for a given snap, so this could probably just return at the first hit instead of returning a []*state.Task

File: overlord/snapstate/handlers.go
Comment (pedronis) [reply overlord/snapstate/handlers.go:364]: this bit should be tested, probably worth checking how something similar is tested for Install

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:417]: there's no test hitting this

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:431]: and this

**Review by pedronis — COMMENTED** [review]
overall looks code, some detail comments

File: overlord/snapstate/conflict.go
Comment (MiguelPires) [reply overlord/snapstate/conflict.go:171]: I wondered about doing it as a separate PR but opted not to since it was only triggered in my branch. I'll split it off into a separate PR with some better testing then

File: overlord/snapstate/handlers.go
Comment (stolowski) [reply overlord/snapstate/handlers.go:218]: +1

File: overlord/snapstate/conflict.go
Comment (MiguelPires) [reply overlord/snapstate/conflict.go:171]: Now that I had another look at this, I think this change is wrong. The issue is just that two updates to the same snap can conflict which means that the second one is retried after 30s while the test expects the task runner to converge sooner than that. I had wondered if blocking the install of the "main" snap because the default provider is outdated was ok. It might be better to significantly shorten the retry period or simply not retry at all, to prevent the "soft" dependency of the content provider from blocking an install for 30 seconds

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:412]: Added tests to fully cover `hasAllContentAttrs`

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go]: The latest commit adds more tests to cover the ways in which conflicts can manifest which covers this path as well

File: overlord/snapstate/conflict.go
Comment (MiguelPires) [reply overlord/snapstate/conflict.go:173]: I haven't change this code in this branch but I opened another PR for this fix (which incorporates this feedback) with proper tests. Once that merges, I'll just rebase this on it and drop this commit

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go]: Thanks, fixed this as well in one of the latest commits

**Comment by MiguelPires** [comment]
Thanks for the feedback. The latest commits expand the tests to fully cover hasAllContentAttrs and the conflicts cases in updatePrereqIfOutdated.  I didn't change the conflict bug fix because I've opened another PR for it so once that merges, this can also merge and I'll just drop the commit from this branch. I also slightly modified the TestUpdateContentProviderDownloadFailure test to not trigger two updates of the same snap just so the retry doesn't delay the task and cause the test to fail. IMO this change is fine since the test's purpose doesn't depend on it, it just depends on the consumer updating the producer

File: overlord/snapstate/handlers.go
Comment (mardy) [inline overlord/snapstate/handlers.go:201]: Maybe rename to `[find]LinkTaskForSnap()`?

I'm not sure what "in flight" means in this context, and whether it's helpful to have it in the function name.

File: overlord/snapstate/handlers.go
Comment (mardy) [inline overlord/snapstate/handlers.go:366]: If it's intentional that we don't return an error, we should have a comment explaining why.

File: overlord/snapstate/conflict.go
Comment (MiguelPires) [reply overlord/snapstate/conflict.go:173]: https://github.com/snapcore/snapd/pull/10767

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:201]: I think it's meant to imply that it only looks for a task that can still run, i.e., a non-Ready() task. But I agree, it's not very clear

**Comment by MiguelPires** [comment]
Thanks everyone for the feedback. I reverted the previous code that dealt with the issue of the DownloadFailure the retry and I did a series of master merges to get the code I needed plus the code to unblock the github actions/unit testing. As for the actual new code, I added the check that skips the update if there's a duplicate one conflicting. I added a test as well to cover this scenario plus a bit of refactoring/clarifying other tests

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:225]: This helper does very little / isn't much different than just using findLinkSnapTask() directly, and appears to be used only once, so maybe just drop it and use findLinkSnapTask?

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:341]: Nitpick^2, this could go before st := ...

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:362]: I wonder if we should maybe reorder the messages here to say that update of prereq failed with err, and therefore it's content is missing?

Comment (MiguelPires) [reply overlord/snapstate/handlers.go:362]: Hm yes, I think that would be better since it'd make the situation clearer for the user. I was trying to imply that there was a prerequisite with missing content and that we tried but failed to update it but that that's not very relevant for the user and might just confuse them

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:338]: I think we are not hitting 'ok' case in the tests?

Comment (MiguelPires) [reply overlord/snapstate/handlers.go:338]: No, because if hard to deterministically hit this case, but there is a `TestHasAllContentAttributes` in `snapstate_install_test.go` that should cover all of it. The issue is that the `ok` case will only be hit when the content is missing before the update tasks are scheduled but is then updated before the handler runs

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:353]: AFAIU (but I may be missing something) we generally avoid conflict with shouldSkipToAvoidConflict() above, so I'm not quite clear what are the exact circumstances we can hit it here (even considering that Update unlocks the state around store call) - but I can see we definitely hit this case in the unit tests. I think both this and the call to shouldSkipToAvoidConflict above should have a comment explaining the conflicts, and why can we hit it again here.

File: overlord/snapstate/handlers.go
Comment (stolowski) [reply overlord/snapstate/handlers.go:338]: Wouldn't this case be hit when the provider snap has all the required content attributes and doesn't need updating, so after Install("consumer") we should see exactly one download happening for the consumer snap and not for the provider?

Comment (MiguelPires) [reply overlord/snapstate/handlers.go:338]: The prerequisites->content map in the prereq task is only populated with content attributes that aren't already provided (I added some comments clarifying that in defaultProviderContentAttrs in an earlier commit). To trigger this case, the content provider can't already have this content at first so the prerequisites task gets a map with the content attributes and default providers to install, but before the prerequisites task of the consumer gets to run the producer is updated. This can happen because an update of the producer would not be prevent since the consumer's prerequisite task hasn't run and the producer install/update tasks haven't been scheduled. I couldn't think of a test that triggers this deterministically

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:353]: It's been a while since the previous check was discussed so I may be forgetting something but, looking at it, it seems like what `shouldSkipToAvoidConflict` does is a subset of the conflict checking in Update so the first could be removed. Maybe @pedronis can correct me if I'm wrong

**Review by mardy — APPROVED** [review]
LGTM, although it's hard to review it properly due to the size. :-)

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:342]: we should drop the TODO, I looked at the code again but Update basically checks for conflicts only after having consulted the store, that's a slow process, so this is just an optimisation in a sense, but I think we do want it

File: overlord/snapstate/handlers.go
Comment (pedronis) [reply overlord/snapstate/handlers.go:355]: Update talks to the store first, so I think this is fine as is

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:355]: I'm not sure it's an easy scenario to reach but I think we want to add the same check to the Install in installOneBaseOrRequired when it checks the ChangeConflictError. Can be done in a small follow-up

**Review by pedronis — APPROVED** [review]
thanks, I haven't looked super closely to the tests, maybe @stolowski can do that.

I think we can drop the TODO about the extra conflict check. I commented on this.

We probably want a follow-up where:

* we add the same change conflict check to the Install path
*  I think because of remodel (this code is older than that), both the Install and the Update
   should actually become InstallWithDeviceContext and UpdateWithDeviceContext using the context from the task via DeviceCtx

File: overlord/snapstate/handlers.go
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:342]: Make sense, I can replace the TODO with a small comment indicating it's an optimization (just to avoid confusion in the future)

File: overlord/snapstate/handlers.go
Comment (stolowski) [reply overlord/snapstate/handlers.go:271]: I think we are not hitting this case in tests.

**Review by stolowski — APPROVED** [review]
Looks good, but I think we're missing a test case for coming from old snapd that didn't populate PrereqContentAttrs? Can be a followup.

**Pull request merged**
